File: backend/models/Product.js
const mongoose = require('mongoose');

// Review schema
const reviewSchema = new mongoose.Schema({
  user: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  rating: { type: Number, required: true, min: 1, max: 5 },
  comment: { type: String, required: true },
  createdAt: { type: Date, default: Date.now }
});

// Product schema
const productSchema = new mongoose.Schema({
  // Basic Info
  name: { type: String, required: true },
  slug: { type: String, unique: true },
  description: { type: String, required: true },
  shortDescription: String,
  // Pricing
  price: { type: Number, required: true },
  originalPrice: Number,
  discountPercentage: { type: Number, default: 0 },
  
  // Images (Up to 3 images + 1 video)
  mainImage: { type: String, required: true },
  images: [String],
  video: String,
  
  // SEO Fields
  seoTitle: String,
  seoDescription: String,
  seoKeywords: [String],
  metaTags: {
    ogTitle: String,
    ogDescription: String,
    ogImage: String,
    twitterCard: String
  },
  
  // Categorization
  category: { type: String, required: true },
  subcategory: String,
  brand: { type: String, default: 'E-Gura Store' },
  tags: [String],
  
  // Product Details
  gender: { type: String, enum: ['male', 'female', 'unisex'], default: 'female' },
  ageGroup: { type: String, enum: ['kids', 'teen', 'adult'], default: 'adult' },
  material: [String],
  care: [String],
  
  // Variants
  sizes: [String],
  colors: [String],
  variants: [{
    size: String,
    color: String,
    price: Number,
    stockQuantity: Number,
    sku: String,
    image: String // Specific image for this variant
  }],
  
  // Inventory
  stockQuantity: { type: Number, default: 0 },
  lowStockThreshold: { type: Number, default: 5 },
  sku: { type: String, unique: true },
  
  // Product Status
  isActive: { type: Boolean, default: true },
  isFeatured: { type: Boolean, default: false },
  isNewProduct: { type: Boolean, default: false }, // renamed from isNew (reserved)
  isSale: { type: Boolean, default: false },
  isBestSeller: { type: Boolean, default: false },
  
  // Reviews and Ratings
  reviews: [reviewSchema],
  averageRating: { type: Number, default: 0 },
  totalReviews: { type: Number, default: 0 },
  
  // SEO and Marketing
  metaTitle: String,
  metaDescription: String,
  keywords: [String],
  
  // Analytics
  viewCount: { type: Number, default: 0 },
  salesCount: { type: Number, default: 0 },
  
  // Shipping
  weight: Number,
  dimensions: {
    length: Number,
    width: Number,
    height: Number
  },
  shippingClass: { type: String, enum: ['standard', 'express', 'free'], default: 'standard' },
  
  // Additional Info
  careInstructions: [String],
  returnPolicy: String,
  warranty: String,
  
  // AI Bargaining Settings
  bargainEnabled: { type: Boolean, default: true },
  minBargainPrice: { type: Number },
  maxBargainDiscount: { type: Number, default: 25, min: 5, max: 50 },
  bargainStrategy: { 
    type: String, 
    enum: ['aggressive', 'balanced', 'conservative'], 
    default: 'balanced' 
  },
  
}, { 
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Virtual for discounted price
productSchema.virtual('discountedPrice').get(function() {
  if (this.discountPercentage > 0) {
    return this.price - (this.price * this.discountPercentage / 100);
  }
  return this.price;
});

// Virtual for stock status
productSchema.virtual('stockStatus').get(function() {
  if (this.stockQuantity === 0) return 'out_of_stock';
  if (this.stockQuantity <= this.lowStockThreshold) return 'low_stock';
  return 'in_stock';
});

// Virtual for calculated average rating
productSchema.virtual('calculatedRating').get(function() {
  if (this.reviews.length === 0) return 0;
  const total = this.reviews.reduce((sum, review) => sum + review.rating, 0);
  return (total / this.reviews.length).toFixed(1);
});

// Pre-save middleware to generate slug
productSchema.pre('save', function(next) {
  if (this.isModified('name')) {
    this.slug = this.name
      .toLowerCase()
      .replace(/[^a-z0-9]+/g, '-')
      .replace(/(^-|-$)/g, '');
  }
  next();
});

// Pre-save middleware to update average rating
productSchema.pre('save', function(next) {
  if (this.reviews.length > 0) {
    const total = this.reviews.reduce((sum, review) => sum + review.rating, 0);
    this.averageRating = (total / this.reviews.length).toFixed(1);
    this.totalReviews = this.reviews.length;
  }
  next();
});

// Index for better search performance
productSchema.index({ name: 'text', description: 'text', tags: 'text' });
productSchema.index({ category: 1, isActive: 1 });
productSchema.index({ isFeatured: 1, isActive: 1 });
productSchema.index({ isNewProduct: 1, isActive: 1 });
productSchema.index({ isSale: 1, isActive: 1 });

module.exports = mongoose.model('Product', productSchema); 

File: mobile/lib/models/product.dart
class Product {
  final String id;
  final String name;
  final double price;
  final double? originalPrice;
  final int? discountPercentage;
  final String? mainImage;
  final String? category;
  final String? subcategory;
  final String? brand;
  final List<String>? tags;
  final List<String>? colors;
  final List<String>? sizes;
  final double? averageRating;
  final int? totalReviews;
  final int? salesCount;
  final int? stockQuantity;
  final bool isFeatured;
  final bool isNew;
  final bool isSale;
  final bool isBestSeller;
  final String? shortDescription;
  final String? description;
  final List<String>? images;
  final String? createdAt;
  bool isFavorite = false;

  Product({
    required this.id,
    required this.name,
    required this.price,
    this.originalPrice,
    this.discountPercentage,
    this.mainImage,
    this.category,
    this.subcategory,
    this.brand,
    this.tags,
    this.colors,
    this.sizes,
    this.averageRating,
    this.totalReviews,
    this.salesCount,
    this.stockQuantity,
    this.isFeatured = false,
    this.isNew = false,
    this.isSale = false,
    this.isBestSeller = false,
    this.shortDescription,
    this.description,
    this.images,
    this.createdAt,
    this.isFavorite = false,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id']?.toString() ?? json['_id']?.toString() ?? '',
      name: json['name']?.toString() ?? 'Unknown Product',
      price: _toDouble(json['price']),
      originalPrice: json['originalPrice'] != null ? _toDouble(json['originalPrice']) : null,
      discountPercentage: _toInt(json['discountPercentage']),
      mainImage: json['mainImage']?.toString(),
      category: json['category']?.toString(),
      subcategory: json['subcategory']?.toString(),
      brand: json['brand']?.toString(),
      tags: _toStringList(json['tags']),
      colors: _toStringList(json['colors']),
      sizes: _toStringList(json['sizes']),
      averageRating: json['averageRating'] != null ? _toDouble(json['averageRating']) : null,
      totalReviews: _toInt(json['totalReviews']),
      salesCount: _toInt(json['salesCount']),
      stockQuantity: json['stockQuantity'] != null ? _toInt(json['stockQuantity']) : null,
      isFeatured: json['isFeatured'] == true,
      isNew: json['isNew'] == true,
      isSale: json['isSale'] == true,
      isBestSeller: json['isBestSeller'] == true,
      shortDescription: json['shortDescription']?.toString(),
      description: json['description']?.toString(),
      images: _toStringList(json['images']),
      createdAt: json['createdAt']?.toString(),
      isFavorite: json['isFavorite'] == true,
    );
  }

  static int _toInt(dynamic value) {
    if (value is int) return value;
    if (value is double) return value.toInt();
    if (value is String) return int.tryParse(value) ?? 0;
    return 0;
  }

  static double _toDouble(dynamic value) {
    if (value is double) return value;
    if (value is int) return value.toDouble();
    if (value is String) return double.tryParse(value) ?? 0.0;
    return 0.0;
  }

  static List<String>? _toStringList(dynamic value) {
    if (value is List) return value.map((e) => e.toString()).toList();
    return null;
  }

  static String formatNumber(double value) {
    final str = value.toStringAsFixed(0);
    final result = StringBuffer();
    for (int i = 0; i < str.length; i++) {
      if (i > 0 && (str.length - i) % 3 == 0) result.write(',');
      result.write(str[i]);
    }
    return result.toString();
  }

  String get formattedPrice => '${formatNumber(price)} RWF';

  String get formattedOriginalPrice =>
      originalPrice != null ? '${formatNumber(originalPrice!)} RWF' : '';

  double get flashDealPrice => (price * 0.7).floorToDouble();

  String get formattedFlashPrice => '${formatNumber(flashDealPrice)} RWF';

  double get savings =>
      originalPrice != null ? originalPrice! - price : price * 0.3;

  String get formattedSavings => '${formatNumber(savings)} RWF';

  bool get hasDiscount => (discountPercentage ?? 0) > 0;

  String get stockStatus {
    if (stockQuantity == null) return 'In Stock';
    if (stockQuantity! <= 0) return 'Out of Stock';
    if (stockQuantity! < 5) return 'Low Stock';
    return 'In Stock';
  }
}

class PaginatedProducts {
  final List<Product> products;
  final int currentPage;
  final int totalPages;
  final int total;

  PaginatedProducts({
    required this.products,
    required this.currentPage,
    required this.totalPages,
    required this.total,
  });

  bool get hasMore => currentPage < totalPages;
}
